vortex-io/src/read_ranges.rs

```diff
@@ -0,0 +1,150 @@
+#![allow(dead_code)]
+use std::future::Future;
+use std::io;
+use std::io::ErrorKind;
+use std::ops::Range;
+use std::sync::Arc;
+
+use bytes::Bytes;
+use futures_util::{stream, FutureExt, StreamExt, TryStreamExt};
+use vortex_error::VortexExpect;
+
+use crate::{Dispatch, IoDispatcher, VortexReadAt};
+
+const MAX_BUFFERED_READS: usize = 10;
+
+#[derive(Debug, Clone)]
+pub struct VortexReadRanges<R> {
+    read: R,
+    dispatcher: Arc<IoDispatcher>,
+    max_gap: usize,
+}
+
+impl<R> VortexReadRanges<R> {
+    pub fn new(read: R, dispatcher: Arc<IoDispatcher>, max_gap: usize) -> VortexReadRanges<R> {
+        Self {
+            read,
+            dispatcher,
+            max_gap,
+        }
+    }
+}
+
+impl<R: VortexReadAt> VortexReadRanges<R> {
```

Comment:
This is closely based on arrow-rs logic but we dispatch each request on our io loop regardless of underlying reader. We shold decide if we require ranges to be sorted by start or will the reader sort them (current implementation)